tests/unit/package_managers/test_npm.py

```diff
@@ -630,3 +715,63 @@ def test_resolve_npm_unsupported_lockfileversion(rooted_tmp_path: RootedPath) ->
     expected_error = f"lockfileVersion {package_lock_json['lockfileVersion']} from {lockfile_path} is not supported"
     with pytest.raises(UnsupportedFeature, match=expected_error):
         _resolve_npm(rooted_tmp_path)
+
+
+@pytest.mark.parametrize(
+    "vcs, expected",
+    [
+        (
+            (
+                "git+ssh://git@bitbucket.org/cachi-testing/cachi2-without-deps.git#9e164b97043a2d91bbeb992f6cc68a3d1015086a"
+            ),
+            {
+                "url": "ssh://git@bitbucket.org/cachi-testing/cachi2-without-deps.git",
+                "ref": "9e164b97043a2d91bbeb992f6cc68a3d1015086a",
+                "host": "bitbucket.org",
+                "namespace": "cachi-testing",
+                "repo": "cachi2-without-deps",
+            },
+        ),
+    ],
+)
+def test_extract_git_info_npm(vcs: str, expected: Dict[str, str]) -> None:
+    assert _extract_git_info_npm(vcs) == expected
+
+
+def test_extract_git_info_with_missing_ref() -> None:
+    vcs = "git+ssh://git@bitbucket.org/cachi-testing/cachi2-without-deps.git"
+    expected_error = (
+        "ssh://git@bitbucket.org/cachi-testing/cachi2-without-deps.git "
+        "is not valid VCS url. ref is missing."
+    )
+    with pytest.raises(UnexpectedFormat, match=expected_error):
+        _extract_git_info_npm(vcs)
+
+
+@pytest.mark.parametrize(
+    "vcs, expected",
+    [
+        ("github:kevva/is-positive#97edff6", "git+ssh://github.com/kevva/is-positive.git#97edff6"),
+        ("github:kevva/is-positive", "git+ssh://github.com/kevva/is-positive.git"),
+        (
+            "bitbucket:cachi-testing/cachi2-without-deps#9e164b9",
+            "git+ssh://bitbucket.org/cachi-testing/cachi2-without-deps.git#9e164b9",
+        ),
+        ("gitlab:foo/bar#YOLO", "git+ssh://gitlab.com/foo/bar.git#YOLO"),
+    ],
+)
+def test_update_vcs_url_with_full_hostname(vcs: str, expected: str) -> None:
+    assert _update_vcs_url_with_full_hostname(vcs) == expected
+
+
+def test_clone_repo_pack_archive(rooted_tmp_path: RootedPath) -> None:
+    vcs = "git+ssh://bitbucket.org/cachi-testing/cachi2-without-deps.git#9e164b9"
+    download_path = _clone_repo_pack_archive(vcs, rooted_tmp_path)
+    expected_path = rooted_tmp_path.join_within_root(
+        "bitbucket.org",
+        "cachi-testing",
+        "cachi2-without-deps",
+        "cachi2-without-deps-external-gitcommit-9e164b9.tgz",
+    )
+    assert download_path == expected_path
+    assert os.path.exists(download_path)
```

Comment:
missing mocks, this unit test actually clones from bitbucket

IMO all that should be tested here is
```python
assert download_path.path.parent.is_dir()
mock_clone_as_tarball.assert_called_once_with(...)
```

Reply:
Fixed